{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Exploration and Cleaning\n",
    "\n",
    "This notebook contains some exploration of the dataset and some cleaning for further analysis. The Readme file for the repository contains information on the data source. \n",
    "\n",
    "Because the CSV is compressed and present in the repository as a XZ file, and read directly compressed with Pandas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we load the data, and explore its structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(138864, 18)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"../data/bq-results-20240802-233415-1722641679845.csv.xz\", \n",
    "                 compression='xz')\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>contig_acc</th>\n",
       "      <th>biosample_acc</th>\n",
       "      <th>bioproject_acc</th>\n",
       "      <th>target_acc</th>\n",
       "      <th>element_symbol</th>\n",
       "      <th>protein_acc</th>\n",
       "      <th>type</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>taxgroup_name</th>\n",
       "      <th>strain</th>\n",
       "      <th>serovar</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>epi_type</th>\n",
       "      <th>host</th>\n",
       "      <th>collection_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AAFUZC010000051.1</td>\n",
       "      <td>SAMN03098832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT000041084.2</td>\n",
       "      <td>blaDHA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CEPHALOSPORIN</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AM49198</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USA</td>\n",
       "      <td>clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AAFUZC010000051.1</td>\n",
       "      <td>SAMN03098832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT000041084.2</td>\n",
       "      <td>ble</td>\n",
       "      <td>EBK1426116.1</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BLEOMYCIN</td>\n",
       "      <td>BLEOMYCIN</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AM49198</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USA</td>\n",
       "      <td>clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AAFUZC010000051.1</td>\n",
       "      <td>SAMN03098832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT000041084.2</td>\n",
       "      <td>blaNDM-1</td>\n",
       "      <td>EBK1426117.1</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CARBAPENEM</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AM49198</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USA</td>\n",
       "      <td>clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AAFUZC010000073.1</td>\n",
       "      <td>SAMN03098832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT000041084.2</td>\n",
       "      <td>blaNDM-1</td>\n",
       "      <td>EBK1426163.1</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CARBAPENEM</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AM49198</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USA</td>\n",
       "      <td>clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AAFUZC010000073.1</td>\n",
       "      <td>SAMN03098832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT000041084.2</td>\n",
       "      <td>ble</td>\n",
       "      <td>EBK1426164.1</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BLEOMYCIN</td>\n",
       "      <td>BLEOMYCIN</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>AM49198</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USA</td>\n",
       "      <td>clinical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       scientific_name         contig_acc biosample_acc bioproject_acc  \\\n",
       "0  Salmonella enterica  AAFUZC010000051.1  SAMN03098832            NaN   \n",
       "1  Salmonella enterica  AAFUZC010000051.1  SAMN03098832            NaN   \n",
       "2  Salmonella enterica  AAFUZC010000051.1  SAMN03098832            NaN   \n",
       "3  Salmonella enterica  AAFUZC010000073.1  SAMN03098832            NaN   \n",
       "4  Salmonella enterica  AAFUZC010000073.1  SAMN03098832            NaN   \n",
       "\n",
       "       target_acc element_symbol   protein_acc type        class  \\\n",
       "0  PDT000041084.2         blaDHA           NaN  AMR  BETA-LACTAM   \n",
       "1  PDT000041084.2            ble  EBK1426116.1  AMR    BLEOMYCIN   \n",
       "2  PDT000041084.2       blaNDM-1  EBK1426117.1  AMR  BETA-LACTAM   \n",
       "3  PDT000041084.2       blaNDM-1  EBK1426163.1  AMR  BETA-LACTAM   \n",
       "4  PDT000041084.2            ble  EBK1426164.1  AMR    BLEOMYCIN   \n",
       "\n",
       "        subclass        taxgroup_name   strain serovar isolation_source  \\\n",
       "0  CEPHALOSPORIN  Salmonella enterica  AM49198     NaN              NaN   \n",
       "1      BLEOMYCIN  Salmonella enterica  AM49198     NaN              NaN   \n",
       "2     CARBAPENEM  Salmonella enterica  AM49198     NaN              NaN   \n",
       "3     CARBAPENEM  Salmonella enterica  AM49198     NaN              NaN   \n",
       "4      BLEOMYCIN  Salmonella enterica  AM49198     NaN              NaN   \n",
       "\n",
       "  geo_loc_name  epi_type host collection_date  \n",
       "0          USA  clinical  NaN             NaN  \n",
       "1          USA  clinical  NaN             NaN  \n",
       "2          USA  clinical  NaN             NaN  \n",
       "3          USA  clinical  NaN             NaN  \n",
       "4          USA  clinical  NaN             NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['scientific_name', 'contig_acc', 'biosample_acc', 'bioproject_acc',\n",
       "       'target_acc', 'element_symbol', 'protein_acc', 'type', 'class',\n",
       "       'subclass', 'taxgroup_name', 'strain', 'serovar', 'isolation_source',\n",
       "       'geo_loc_name', 'epi_type', 'host', 'collection_date'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The dataset has 138,886 rows and 18 columns. The description of the columns is the following:\n",
    "\n",
    "- scientific_name: the bacterial species name\n",
    "- contig_acc: the contig accesion number from NCBI, where the gene is present\n",
    "- biosample_acc: the biosample accession number\n",
    "- bioproject_acc: the bioproject accession number\n",
    "- target_acc: unsure, this seems to be an accession number for the genome assembly version\n",
    "- element_symbol: the gene name (e.g. _blaNDM-1_)\n",
    "- protein_acc: the protein accession number for the gene\n",
    "- type: the type of mechanism (e.g. AMR, antimicrobial resistance)\n",
    "- class: the functional class for the mechanism in this row\n",
    "- subclass: the functional subclass for the mechanism in this row\n",
    "- taxgroup_name: The taxonomic group. It may be similar to the scientific name, but unsure at the moment\n",
    "- strain: the strain for this genome\n",
    "- serovar: For some taxa, it could be a serovar classification.\n",
    "- isolation_source: The biological source of the sample (it seems). For example, blood.\n",
    "- geo_loc_name: the geographical origin of the isolate\n",
    "- epi_type: needs some exploration, it seems to be the type of strain (clinical, maybe environmental?)\n",
    "- host: The organism where this bacteria was isolated, for example _Homo sapiens_ (human)\n",
    "- collection_date: The date where the isolate was collected.\n",
    "\n",
    "Just from the first five rows, we can see that not all the columns are complete, and we need some additional filtering to have the data ready for analysis.\n",
    "\n",
    "We can look in more detail the dataset, and use this information for further cleaning, including a description of the dataset, and looking at the NA values present in it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>contig_acc</th>\n",
       "      <th>biosample_acc</th>\n",
       "      <th>bioproject_acc</th>\n",
       "      <th>target_acc</th>\n",
       "      <th>element_symbol</th>\n",
       "      <th>protein_acc</th>\n",
       "      <th>type</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>taxgroup_name</th>\n",
       "      <th>strain</th>\n",
       "      <th>serovar</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>epi_type</th>\n",
       "      <th>host</th>\n",
       "      <th>collection_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>138864</td>\n",
       "      <td>138864</td>\n",
       "      <td>138864</td>\n",
       "      <td>26693</td>\n",
       "      <td>138864</td>\n",
       "      <td>138864</td>\n",
       "      <td>121804</td>\n",
       "      <td>138864</td>\n",
       "      <td>138007</td>\n",
       "      <td>138007</td>\n",
       "      <td>138864</td>\n",
       "      <td>98238</td>\n",
       "      <td>1983</td>\n",
       "      <td>104426</td>\n",
       "      <td>136244</td>\n",
       "      <td>124544</td>\n",
       "      <td>122637</td>\n",
       "      <td>132082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>96</td>\n",
       "      <td>41640</td>\n",
       "      <td>26809</td>\n",
       "      <td>692</td>\n",
       "      <td>26891</td>\n",
       "      <td>525</td>\n",
       "      <td>84585</td>\n",
       "      <td>3</td>\n",
       "      <td>37</td>\n",
       "      <td>64</td>\n",
       "      <td>32</td>\n",
       "      <td>18425</td>\n",
       "      <td>71</td>\n",
       "      <td>1268</td>\n",
       "      <td>907</td>\n",
       "      <td>2</td>\n",
       "      <td>107</td>\n",
       "      <td>3570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NZ_CP095443.1</td>\n",
       "      <td>SAMN27484493</td>\n",
       "      <td>PRJNA288601</td>\n",
       "      <td>PDT001315396.1</td>\n",
       "      <td>ble</td>\n",
       "      <td>WP_004201167.1</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CARBAPENEM</td>\n",
       "      <td>E.coli and Shigella</td>\n",
       "      <td>308</td>\n",
       "      <td>not available</td>\n",
       "      <td>urine</td>\n",
       "      <td>USA</td>\n",
       "      <td>clinical</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>58548</td>\n",
       "      <td>109</td>\n",
       "      <td>129</td>\n",
       "      <td>5283</td>\n",
       "      <td>129</td>\n",
       "      <td>40865</td>\n",
       "      <td>7959</td>\n",
       "      <td>123952</td>\n",
       "      <td>46061</td>\n",
       "      <td>41690</td>\n",
       "      <td>58667</td>\n",
       "      <td>776</td>\n",
       "      <td>447</td>\n",
       "      <td>25639</td>\n",
       "      <td>40911</td>\n",
       "      <td>118880</td>\n",
       "      <td>116559</td>\n",
       "      <td>11480</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         scientific_name     contig_acc biosample_acc bioproject_acc  \\\n",
       "count             138864         138864        138864          26693   \n",
       "unique                96          41640         26809            692   \n",
       "top     Escherichia coli  NZ_CP095443.1  SAMN27484493    PRJNA288601   \n",
       "freq               58548            109           129           5283   \n",
       "\n",
       "            target_acc element_symbol     protein_acc    type        class  \\\n",
       "count           138864         138864          121804  138864       138007   \n",
       "unique           26891            525           84585       3           37   \n",
       "top     PDT001315396.1            ble  WP_004201167.1     AMR  BETA-LACTAM   \n",
       "freq               129          40865            7959  123952        46061   \n",
       "\n",
       "          subclass        taxgroup_name strain        serovar  \\\n",
       "count       138007               138864  98238           1983   \n",
       "unique          64                   32  18425             71   \n",
       "top     CARBAPENEM  E.coli and Shigella    308  not available   \n",
       "freq         41690                58667    776            447   \n",
       "\n",
       "       isolation_source geo_loc_name  epi_type          host collection_date  \n",
       "count            104426       136244    124544        122637          132082  \n",
       "unique             1268          907         2           107            3570  \n",
       "top               urine          USA  clinical  Homo sapiens            2023  \n",
       "freq              25639        40911    118880        116559           11480  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe(include=\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scientific_name          0\n",
       "contig_acc               0\n",
       "biosample_acc            0\n",
       "bioproject_acc      112171\n",
       "target_acc               0\n",
       "element_symbol           0\n",
       "protein_acc          17060\n",
       "type                     0\n",
       "class                  857\n",
       "subclass               857\n",
       "taxgroup_name            0\n",
       "strain               40626\n",
       "serovar             136881\n",
       "isolation_source     34438\n",
       "geo_loc_name          2620\n",
       "epi_type             14320\n",
       "host                 16227\n",
       "collection_date       6782\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One of the first questions is to find a unique identifier that can be used across the dataset. Each entry may have multiple rows associated, and variables such as `scientific_name` can be found multiple times. Looking at the description of the dataset some things that come up to attention:\n",
    "\n",
    "- biosample_acc: This appears to be the best identifier. From NCBI, this refers is a unique identifier associated with the biological sample. There are no null values in this column.\n",
    "- bioproject_acc: Not present in all of the rows. Not a good identifier.\n",
    "- target_acc: according to the [NCBI documentation](https://www.ncbi.nlm.nih.gov/pathogens/pathogens_help/#isolates-browser-data-field-Isolate) it refers to the pathogen detection accession of the isolate. Probably could be used as an ID, but the number is higher compared to the biosample_acc. Maybe there is more than one genome assembly per biosample, this needs to be further explored.\n",
    "- For further analysis and some of the questions, not all samples have isolation_source, geo_loc_name, host, and date, so we have to keep this in mind."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
